File: 01-intro/ruby/binary_search.rb
# frozen_string_literal: true

def binary_search(list, item)
  low = 0
  high = list.size - 1

  loop do
    break if low > high

    mid = (high + low) / 2
    guess = list[mid]
    if guess == item
      return mid
    elsif guess < item
      low = mid + 1
    else
      high = mid - 1
    end
  end
end

File: 01-intro/python/binary_search.py
def binary_search(sorted_list, item):
    """Implementation of binary search"""
    low = 0
    high = len(sorted_list) - 1

    while low <= high:
        mid = round((high + low) / 2)
        guess = sorted_list[mid]

        if guess == item:
            return mid
        if guess < item:
            low = mid + 1
        else:
            high = mid - 1
    return None
